File: java/0234_isPalindrome.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        List<Integer> arr = new ArrayList<>();
        while(head!=null)
        {
            arr.add(head.val);
            head = head.next;
        }
        int p1=0,p2=(int)arr.size()-1;
        while(p1<=p2)
        {
            if(arr.get(p1)!=arr.get(p2))
                return false;
            p1++;p2--;
        }
        return true;
    }
}

File: cpp/0234_isPalindrome.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// 用栈或者数组，双指针shrink检验回文
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> arr;
        while(head!=nullptr)
        {
            arr.push_back(head->val);
            head = head->next;
        }
        auto p1=arr.begin(),p2=arr.end()-1;
        while(p1<=p2)
        {
            if(*p1!=*p2)
                return false;
            p1++;p2--;
        }
        return true;
    }
};
